extensions/amp-video/1.0/video-iframe.js

```diff
@@ -0,0 +1,159 @@
+  useMemo,
+  useRef,
+} from '../../../src/preact';
+
+const defaultSandbox = [
+  'allow-scripts',
+  'allow-same-origin',
+  'allow-popups',
+  'allow-popups-to-escape-sandbox',
+  'allow-top-navigation-by-user-activation',
+].join(' ');
+
+/**
+ * Goes inside a VideoWrapper.
+ *
+ *    import {VideoIframe} from '.../video-iframe';
+ *    import {VideoWrapper} from '.../video-wrapper';
+ *    render(<VideoWrapper component={VideoIframe} ... />)
+ *
+ * Usable on the AMP layer through VideoBaseElement.
+ * @param {!VideoIframeDef.Props} props
+ * @param {{current: (T|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function VideoIframeWithRef(
+  {
+    sandbox = defaultSandbox,
+    muted = false,
+    controls = false,
+    onCanPlay,
+    onMessage,
+    sendMessage,
+    ...rest
+  },
+  ref
+) {
+  const iframeRef = useRef(null);
+
+  const readyDeferred = useMemo(() => new Deferred(), []);
+
+  const postMessage = useCallback(
+    (message) => {
+      if (!iframeRef.current || !iframeRef.current.contentWindow) {
+        return;
+      }
+      readyDeferred.promise.then(() => {
+        const transformed = sendMessage(message);
+        iframeRef.current.contentWindow./*OK*/ postMessage(
+          typeof transformed == 'string'
+            ? transformed
+            : JSON.stringify(transformed),
+          '*'
+        );
+      });
+    },
+    [readyDeferred.promise, sendMessage]
+  );
+
+  useImperativeHandle(
+    ref,
+    () => {
+      return {
+        play: () => postMessage('play'),
+        pause: () => postMessage('pause'),
+      };
+    },
+    [postMessage]
+  );
+
+  useLayoutEffect(() => {
+    if (!onMessage) {
+      return;
+    }
+
+    /** @param {Event} event */
+    function handleMessage(event) {
+      if (
+        // handle origin?
+        // event.origin != 'https://www.instagram.com' ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      // Triggers like an HTMLMediaElement, so we give it an iframe handle
+      // to dispatch events from. They're caught from being set on {...rest} so
+      // setting onPlay, etc. props should just work.
+      onMessage({
+        data: event.data,
+        currentTarget: iframeRef.current,
+      });
+    }
+
+    const {defaultView} = iframeRef.current.ownerDocument;
+
+    defaultView.addEventListener('message', handleMessage);
+
+    return () => defaultView.removeEventListener('message', handleMessage);
+  }, [onMessage]);
+
+  useLayoutEffect(() => {
+    if (muted) {
+      postMessage('mute');
+    } else {
+      postMessage('unmute');
+    }
+  }, [muted, postMessage]);
+
+  useLayoutEffect(() => {
+    if (controls) {
+      postMessage('showControls');
+    } else {
+      postMessage('hideControls');
+    }
+  }, [controls, postMessage]);
+
+  return (
+    <iframe
+      {...rest}
+      ref={iframeRef}
+      allowfullscreen
+      frameborder="0"
+      sandbox={sandbox}
+      onCanPlay={() => {
```

Comment:
Is `iframe.oncanplay` a thing? Or is that the result of our own dispatch?

Reply:
We dispatch it ourselves.